### PR TITLE
fix(tokens): graceful tiktoken fallback for local/airgapped endpoints

### PR DESCRIPTION
**Pull request by TimeToBuildBob**
## Summary

- Wraps tiktoken BPE data fetch in a `ThreadPoolExecutor` with a configurable timeout (`GPTME_TIKTOKEN_TIMEOUT`, default 5 s) so `gptme` never blocks indefinitely when encoding files aren't cached and the host can't reach `openaipublic.blob.core.windows.net`
- Fixes the common local-model setup pain: Ollama / vLLM / LiteLLM / airgapped systems hang on startup because tiktoken tries to download `cl100k_base.tiktoken` (~1 MB) on first use
- Strips provider prefixes (`openai/`, `anthropic/`, …) before the `MODEL_TO_ENCODING` static-dict lookup so `openai/o1` → `o200k_base` instead of falling through to `cl100k_base`

## Behaviour

| Scenario | Before | After |
|---|---|---|
| tiktoken cached (normal) | works | works (unchanged) |
| tiktoken not installed | returns `None` | returns `None` (unchanged) |
| BPE fetch hangs (airgapped) | **hangs indefinitely** | logs warning, returns `None`, uses `~4 chars/token` |
| `GPTME_TIKTOKEN_TIMEOUT=0` | N/A | skips tiktoken entirely from the start |
| `openai/o1` model | → `cl100k_base` (wrong) | → `o200k_base` (correct) |

The `len_tokens()` signature is unchanged. Approximated counts are not stored in `_token_cache` so accurate counts are used if the tokenizer later becomes available.

## Configuration

```bash
# Default — 5 second timeout for BPE data download
# Set higher on slow connections:
export GPTME_TIKTOKEN_TIMEOUT=30

# Disable tiktoken entirely (always use char-based approximation):
export GPTME_TIKTOKEN_TIMEOUT=0

# Pre-cache encodings for offline use:
python -c "import tiktoken; tiktoken.get_encoding('cl100k_base'); tiktoken.get_encoding('o200k_base')"
# Or point at a directory with pre-downloaded .tiktoken files:
export TIKTOKEN_CACHE_DIR=/path/to/cache
```

## Test plan

- [x] `test_get_tokenizer_timeout_fallback` — simulates a hanging download, asserts `None` returned
- [x] `test_get_tokenizer_disabled_by_timeout_zero` — asserts `_load_encoding` never called when `TIMEOUT=0`
- [x] `test_len_tokens_fallback_approximation` — asserts `400 chars → 100 tokens` via char fallback
- [x] `test_len_tokens_approximation_not_cached` — asserts approximated results are not stored in cache
- [x] All 21 tests pass: `uv run --with tiktoken pytest tests/test_util_tokens.py`

### CODE REVIEW

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
Addressed the three P1 findings from Greptile's review:

**1. Bad env value crash** (`float()` at module import)
`_parse_tiktoken_timeout()` now wraps the env parse in try/except ValueError, falling back to 5.0 with a log warning. Tests: `test_tiktoken_timeout_invalid_env`, `test_tiktoken_timeout_valid_env`.

**2. `@lru_cache` cached `None` (timeout/failure) permanently**
Replaced with `_GetTokenizer` callable class + `_tokenizer_cache` dict that only stores successful loads. Timed-out or failed calls are not cached, so a retry after network recovery (or after `TIKTOKEN_CACHE_DIR` is populated) works correctly. `cache_clear()` is now a proper method on the class — no dynamic attribute attachment needed. Tests: `test_get_tokenizer_timeout_not_cached`, `test_get_tokenizer_successful_load_is_cached`.

**3. Non-daemon thread kept short-lived CLI processes alive**
`ThreadPoolExecutor` replaced with `threading.Thread(daemon=True)` — a timed-out fetch thread will no longer block process exit.

All 25 tests pass, mypy clean, pre-commit clean.

File: gptme/util/tokens.py [hunk 1]
```diff
@@ -12,59 +13,187 @@
...
+            # which can delay CLI exit by the full TCP timeout.
+            result_holder: list[tiktoken.Encoding | Exception | None] = [None]
+
+            def _do_load() -> None:
+                try:
+                    result_holder[0] = _load_encoding(encoding_name)
+                except Exception as exc:
+                    result_holder[0] = exc
+
+            t = threading.Thread(target=_do_load, daemon=True, name="tiktoken-fetch")
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/util/tokens.py]: Fixed in 1892f64af — added `_inflight_loads: dict[str, tuple[Thread, list]]` (protected by `_inflight_lock`) at module level. New callers for an encoding that already has a thread in flight join that thread via `t.join(timeout=…)` instead of spawning a duplicate. The thread removes itself from `_inflight_loads` on completion (success or error) via a `finally` block, so retries after network recovery still work. Added `test_get_tokenizer_no_duplicate_threads` which asserts exactly one `_load_encoding` call across two sequential timed-out calls.

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
Re-reviewed after the thread-dedup fix (1892f64af). Greptile remains at 4/5 with two remaining concerns:

1. **Non-finite timeout values** — `_parse_tiktoken_timeout` accepts `float('nan')` without error; `t.join(timeout=nan)` has undefined behavior in CPython. Fix: add `math.isnan(val)` guard alongside the existing `ValueError` catch.

2. **Permanently-stuck in-flight worker** — if the download thread never exits (rare: OS TCP never times out), all subsequent calls for that encoding join the same blocked thread and return `None` indefinitely. This is a known trade-off of the in-flight dedup approach vs. the previous behavior (which spawned a new thread each time but otherwise had the same outcome). A mitigation would be to also track a `threading.Event` per encoding that gets set when TIKTOKEN_CACHE_DIR becomes usable, allowing waiting callers to break out.

Stopping here per the one-re-review rule. The self-merge gate correctly blocks at 4/5. Leaving this for human review/decision on whether to address these edge cases.

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
## Status Update

✅ **Greptile Review: 5/5 (safe to merge)**

All code feedback addressed across 5 monitoring sessions:
- **P1 fixes**: Bad env value crash, cached None on timeout, daemon thread blocking (all fixed)
- **Edge cases**: NaN/infinite timeout values guarded
- **Thread dedup**: In-flight encoding loads reused to prevent thread spam on timeouts

**CI Status**:
- ✅ Lint, typecheck, docker, build, benchmark: all pass
- ✅ API tests (openai, openrouter/anthropic): all pass
- ⏳ 2 test jobs still in progress after 4+ min ("Test without API keys" variants — may be slow or hit a known flake; logs will be available on completion)

**Coverage**: 68% patch coverage; 50 lines in tests missing coverage (expected for timeout/edge-case testing).

Ready for your review and merge. Will update if CI jobs surface any issues.

**Comment by TimeToBuildBob** [comment]
CI complete. Summary:

- ✅ Greptile: 5/5 (safe to merge)
- ✅ lint, typecheck, build, Docker, Startup Benchmark, API tests (openai + haiku)
- ✅ Test without API keys or extras (9m)
- ❌ Test without API keys — **pre-existing flake, not caused by this PR**

The failure is `test_e2e_queue_dispatches_after_turn` with a pytest stash `KeyError` in teardown. The test body itself **passed on attempt 2** (retried successfully). The error is a pytest fixture teardown race, not a tiktoken regression. This PR only changes the tokenizer fallback path in `gptme/llm/llm_openai.py` + tests.

Ready for merge. Tagging @ErikBjare for the final call.
